pkg/solana/logpoller/loader.go

```diff
@@ -0,0 +1,287 @@
+package logpoller
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+type ProgramEventProcessor interface {
+	// Process should take a ProgramEvent and parseProgramLogs it based on log signature
+	// and expected encoding. Only return errors that cannot be handled and
+	// should exit further transaction processing on the running thread.
+	//
+	// Process should be thread safe.
+	Process(ProgramEvent) error
+}
+
+type RPCClient interface {
+	GetLatestBlockhash(ctx context.Context, commitment rpc.CommitmentType) (out *rpc.GetLatestBlockhashResult, err error)
+	GetBlocks(ctx context.Context, startSlot uint64, endSlot *uint64, commitment rpc.CommitmentType) (out rpc.BlocksResult, err error)
+	GetBlockWithOpts(context.Context, uint64, *rpc.GetBlockOpts) (*rpc.GetBlockResult, error)
+	GetSignaturesForAddressWithOpts(context.Context, solana.PublicKey, *rpc.GetSignaturesForAddressOpts) ([]*rpc.TransactionSignature, error)
+}
+
+const (
+	DefaultNextSlotPollingInterval = 1_000 * time.Millisecond
+)
+
+type EncodedLogCollector struct {
+	// service state management
+	services.Service
+	engine *services.Engine
+
+	// dependencies and configuration
+	client       RPCClient
+	parser       ProgramEventProcessor
+	lggr         logger.Logger
+	rpcTimeLimit time.Duration
+
+	// internal state
+	chSlot  chan uint64
+	chBlock chan uint64
+	chJobs  chan Job
+	workers *WorkerGroup
+
+	highestSlot       atomic.Uint64
+	highestSlotLoaded atomic.Uint64
+	lastSentSlot      atomic.Uint64
+}
+
+func NewEncodedLogCollector(
+	client RPCClient,
+	parser ProgramEventProcessor,
+	lggr logger.Logger,
+) *EncodedLogCollector {
+	c := &EncodedLogCollector{
+		client:       client,
+		parser:       parser,
+		chSlot:       make(chan uint64),
+		chBlock:      make(chan uint64, 1),
+		chJobs:       make(chan Job, 1),
+		lggr:         lggr,
+		rpcTimeLimit: 1 * time.Second,
+	}
+
+	c.Service, c.engine = services.Config{
+		Name: "EncodedLogCollector",
+		NewSubServices: func(lggr logger.Logger) []services.Service {
+			c.workers = NewWorkerGroup(DefaultWorkerCount, lggr)
+
+			return []services.Service{c.workers}
+		},
+		Start: c.start,
+		Close: c.close,
+	}.NewServiceEngine(lggr)
+
+	return c
+}
+
+func (c *EncodedLogCollector) BackfillForAddress(ctx context.Context, address string, fromSlot uint64) error {
+	pubKey, err := solana.PublicKeyFromBase58(address)
+	if err != nil {
+		return err
+	}
+
+	var (
+		lowestSlotRead uint64
+		lowestSlotSig  solana.Signature
+	)
+
+	for lowestSlotRead > fromSlot || lowestSlotRead == 0 {
+		opts := rpc.GetSignaturesForAddressOpts{
+			Commitment:     rpc.CommitmentFinalized,
+			MinContextSlot: &fromSlot,
```

Comment:
```suggestion
			MinContextSlot: &toSlot,
```
We might need to introduce toSlot to guarantee that backfill and forward processing do not have gaps for a new filter

Reply:
Are you meaning a range filter with both from and to? The RPC call is weird because the from endpoint is a slot number and the to endpoint is a transaction signature.

Reply:
I didn't realize this until after leaving mine, but I think he might be saying the same thing or similar to my comment above.  LogPoller must always process backfills in ascending order of block number, otherwise gaps can be introduced by node restarts which can result in logs getting permanently missed.

Reply:
AFAIK the `GetSignaturesForAddressOpts` does not allow to go forward (from oldest to newest), so the only way to avoid gaps is to sync backfill and forward processing to ensure backfill starts going backwards from a block that was already processed by forward processing.

Reply:
@dhaidashenko I left a suggestion here for how we could address this:
https://github.com/smartcontractkit/chainlink-solana/pull/911#discussion_r1847593321

I guess it is not using the api for what it was designed for tho.  If we're only going to have 24 hours of history, then maybe it's not worth worrying about, we can just reverse it in the next layer. I'd mostly be worried if we had to go back a long ways like we do on evm.

Reply:
But we can't specify a `to` block/slot. We can only specify a transaction signature that acts as the upper limit. Maybe we will need this, but maybe not. I think we can add it later if we end up needing it though.

Reply:
Even if we could have a `to_slot` I don't follow @dhaidashenko 's reasoning about how/why that could help avoid gaps.

If we eventually implement a replay feature, it could be helpful for skipping over blocks that we've already pulled (to get to earlier ones), but it can't ever give us extra results that would be missing without it.

For now, the only use case we'll have is fetching new results up through and including the current block--where fromBlock will be passed in by `RegisterFilter` the first time, and then set to the most recent block in the db each additional time it's called.

That reminds me, is there an easy way to convert a `fromBlock` to a `fromSlot`? I guess we might have to call an extra rpc call to that.

Reply:
> If we eventually implement a replay feature, it could be helpful for skipping over blocks that we've already pulled (to get to earlier ones)

Also: when we do get around to implementing Replay, I think looking up the earliest log in the db and passing that transaction hash as the upper limiting signature will if anything be easier than it would be if we were able to instead use the earliest block in the db for that filter and use that to compute a `toSlot` to pass.  At least in the previous case, we don't have to do any conversion.